File: api/storage/azure_blob.py
"""Azure Blob Storage backend – equivalent to the TypeScript AzureBlobAdapter."""

from __future__ import annotations

import json
from datetime import datetime, timedelta, timezone

from azure.storage.blob import (
    BlobSasPermissions,
    BlobServiceClient,
    generate_blob_sas,
)

from models import Decision, DecisionMeta, to_meta
from .base import StorageBackend


class AzureBlobBackend(StorageBackend):
    def __init__(self, connection_string: str, container: str) -> None:
        self._service = BlobServiceClient.from_connection_string(connection_string)
        self._container_name = container
        # Ensure the container exists on startup
        container_client = self._service.get_container_client(self._container_name)
        if not container_client.exists():
            container_client.create_container()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    def _blob_name(self, prefix: str, decision_id: str) -> str:
        return f"{prefix}{decision_id}.json"

    @property
    def _container(self):
        return self._service.get_container_client(self._container_name)

    def _blob(self, prefix: str, decision_id: str):
        return self._container.get_blob_client(self._blob_name(prefix, decision_id))

    # ------------------------------------------------------------------
    # Interface
    # ------------------------------------------------------------------

    async def list(self, prefix: str) -> list[DecisionMeta]:
        metas: list[DecisionMeta] = []
        for blob_props in self._container.list_blobs(name_starts_with=prefix):
            if not blob_props.name.endswith(".json"):
                continue
            try:
                decision_id = blob_props.name.removeprefix(prefix).removesuffix(".json")
                decision = await self.get(prefix, decision_id)
                metas.append(to_meta(decision))
            except Exception:
                # skip malformed blobs, matching TS behaviour
                continue

        metas.sort(key=lambda m: m.updated_at, reverse=True)
        return metas

    async def get(self, prefix: str, decision_id: str) -> Decision:
        blob_client = self._blob(prefix, decision_id)
        data = blob_client.download_blob().readall()
        return Decision.model_validate_json(data)

    async def put(self, prefix: str, decision_id: str, data: Decision) -> None:
        blob_client = self._blob(prefix, decision_id)
        body = json.dumps(data.model_dump(by_alias=True), indent=2)
        blob_client.upload_blob(body, overwrite=True, content_settings=_json_content())

    async def delete(self, prefix: str, decision_id: str) -> None:
        blob_client = self._blob(prefix, decision_id)
        blob_client.delete_blob(delete_snapshots="include")

    async def share(self, prefix: str, decision_id: str, email: str | None = None) -> str:
        blob_client = self._blob(prefix, decision_id)

        # Try generating a SAS token
        account_name = self._service.account_name
        account_key: str | None = None
        if self._service.credential and hasattr(self._service.credential, "account_key"):
            account_key = self._service.credential.account_key  # type: ignore[attr-defined]

        if account_name and account_key:
            sas_token = generate_blob_sas(
                account_name=account_name,
                container_name=self._container_name,
                blob_name=self._blob_name(prefix, decision_id),
                account_key=account_key,
                permission=BlobSasPermissions(read=True),
                expiry=datetime.now(timezone.utc) + timedelta(days=7),
            )
            return f"{blob_client.url}?{sas_token}"

        # Fall back to the plain blob URL
        return blob_client.url


def _json_content():
    from azure.storage.blob import ContentSettings

    return ContentSettings(content_type="application/json")

File: src/lib/storage/azure-blob.ts
import {
  BlobServiceClient,
  generateBlobSASQueryParameters,
  BlobSASPermissions,
  StorageSharedKeyCredential,
} from "@azure/storage-blob"
import type { StorageAdapter } from "./types"
import type { Decision, DecisionMeta } from "@/lib/decisions/schema"
import { toMeta } from "@/lib/decisions/schema"

export interface AzureConfig {
  connectionString: string
  container: string
  prefix?: string
}

export class AzureBlobAdapter implements StorageAdapter {
  private serviceClient: BlobServiceClient
  private container: string
  private prefix: string

  constructor(config: AzureConfig) {
    this.serviceClient = BlobServiceClient.fromConnectionString(
      config.connectionString,
    )
    this.container = config.container
    this.prefix = config.prefix ?? "decisions/"
  }

  private blobName(id: string) {
    return `${this.prefix}${id}.json`
  }

  private containerClient() {
    return this.serviceClient.getContainerClient(this.container)
  }

  async list(): Promise<DecisionMeta[]> {
    const client = this.containerClient()
    const metas: DecisionMeta[] = []

    for await (const blob of client.listBlobsFlat({ prefix: this.prefix })) {
      if (!blob.name.endsWith(".json")) continue
      try {
        const id = blob.name.replace(this.prefix, "").replace(".json", "")
        const decision = await this.get(id)
        metas.push(toMeta(decision))
      } catch {
        // skip malformed
      }
    }

    return metas.sort(
      (a, b) =>
        new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime(),
    )
  }

  async get(id: string): Promise<Decision> {
    const blob = this.containerClient().getBlockBlobClient(this.blobName(id))
    const res = await blob.download()
    const body = await streamToString(res.readableStreamBody!)
    return JSON.parse(body)
  }

  async put(id: string, data: Decision): Promise<void> {
    const blob = this.containerClient().getBlockBlobClient(this.blobName(id))
    const body = JSON.stringify(data, null, 2)
    await blob.upload(body, Buffer.byteLength(body), {
      blobHTTPHeaders: { blobContentType: "application/json" },
    })
  }

  async delete(id: string): Promise<void> {
    const blob = this.containerClient().getBlockBlobClient(this.blobName(id))
    await blob.deleteIfExists()
  }

  async share(id: string): Promise<string> {
    const client = this.containerClient()
    const blob = client.getBlockBlobClient(this.blobName(id))

    // Extract account name and key from connection string for SAS generation
    const accountMatch = this.serviceClient.accountName
    const cred = this.serviceClient.credential

    if (!(cred instanceof StorageSharedKeyCredential)) {
      // Fall back to blob URL without SAS if credential type doesn't support it
      return blob.url
    }

    const expiry = new Date()
    expiry.setDate(expiry.getDate() + 7)

    const sas = generateBlobSASQueryParameters(
      {
        containerName: this.container,
        blobName: this.blobName(id),
        permissions: BlobSASPermissions.parse("r"),
        expiresOn: expiry,
      },
      cred,
    )

    return `${blob.url}?${sas.toString()}`
  }
}

async function streamToString(
  readable: NodeJS.ReadableStream,
): Promise<string> {
  const chunks: Buffer[] = []
  for await (const chunk of readable) {
    chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk as string))
  }
  return Buffer.concat(chunks).toString("utf8")
}
